readme.txt
Ruined

Ruined is a browser base debugger, using pure ruby script and javascript.

flash demo:
 http://www.artonx.org/data/ruined/

features:
 - step
 - auto run
 - break
 - local variables viewer
 - global variables viewer
 - self's instance variables viewer
 - stdout viewer
 - change variable's value
 - editing source and save (save & restart, save & continue, discard change)

 todo:
  - more inspection 
  - start with none-existing file (using shell or bat)

usage:

simply add '-ruined' for your command line.
cf)
ruby -ruined target.rb [ target's options ...]

licence:
 Simplified BSDL or (GPL3 or any later version)

others:
 jQuery (bundled)

author:
 arton (artonx@gmail.com)


ChangeLog
Tue Nov 9 2:42:30 2010 arton
 * ruined/ruinmain.rb
	write file with binary mode for voiding bad conversion (\r\n -> \r\r\n)
 * uined.rb
	return waitcline.html if restart
 * ruined/html/main.html
	implements edit command
Mon Oct 25 22:41:20 2010 arton
 * ruined/ruinmain.rb
	version 0.0.9 => 0.1.0
	convert file, variable's value and output into utf-8
Sun Oct 17 19:24:00 2010 arton	
 * ruined/ruinmain.rb
	version 0.0.8 => 0.0.9
	correct thread status (exchange nil and false)
Sun Oct 17 19:09:00 2010 arton
 * ruined/ruinmain.rb
	version 0.0.7 => 0.0.8	
	retuens thread-list
 * ruined/css/ruin.css	
	move style from main.html to here
 * ruined/html/main.html
	add thread list pane.
	adjust some styles.
Thu Oct 14 20:14:00 2010 arton	
 * ruined/ruinmain.rb
	version 0.0.5 => 0.0.7
	fix caller check
Wed Oct 13 22:39:00 2010 arton
 * ruined/ruinmain.rb
	version 0.0.5 => 0.0.6
	support multi-thread app (but very limited)
Wed Oct 13 21:01:00 2010 arton
 * Rakefile
	add BSDL into gem
 * ruined/html/main.html
	escape modified value
 * ruined/ruinmain.rb
	version 0.0.4 => 0.0.5
	escape stdout
	unescape modified value
Tue Oct 12 00:17:00 2010 arton	
 * BSDL
 * readme.txt
	license changed to dual (BSDL or (GPLv3 or later))
 * ruined/html/main.html
	support variable modification
 * ruined/ruinmain.rb
	support variable modification
Mon Oct 11 21:25:00 2010 arton		
 * lib/ruined/ruinmain.rb
	fix local vairable confliction with 'v'
	change version => 0.0.3 for new gem
Mon Oct 11 18:38:00 2010 arton	
 * lib/ruined/ruinmain.rb
	change version => 0.0.2 for new gem
Mon Oct 11 15:51:00 2010 arton
	start brwoser (suggested by knu and n0kada, thanks)
Mon Oct 11 01:37:00 2010 arton
	initial release


lib/test.rb
require_relative 'test2.rb'

a = 32
b = a * 5
puts 'hello'
puts b.to_s


test/test.rb
#!/usr/local/ruby -Ku
# coding: utf-8
# $Id:$

require 'test/unit'
require 'webrick'

class Test
  def test_none
    assert true
  end
end


lib/uined.rb
#!/usr/local/bin/ruby -Ku
# coding: utf-8

require 'webrick/log'

def run_app()
  argv = $DEBUG ? ['-d'] : []
  argv += ['-rruined/ruinmain', $0]
  argv += ARGV
  spawn "#{RbConfig::ruby}", *argv
end

def kill_child()
  open('http://localhost:8383/quit') do |h|
    h.read
  end
end

def quit_svr(c, svr)
  Thread.start do
    sleep(1)
    svr.shutdown
    puts 'svr stopped'
  end
end

if $0 == __FILE__
  $stderr.puts 'usage: ruby -ruined target [target-args]'
else
  require 'webrick'
  require 'open-uri'
  require 'mkmf'

  include WEBrick
  svr = HTTPServer.new(:Port => 8384,
                       :DocumentRoot => "#{File.dirname(__FILE__)}/ruined")
  trap('INT') do 
    svr.shutdown
  end
  svr.mount_proc('/restart') do |req, res|
    begin
      kill_child
    rescue
      svr.logger.error 'failed to kill child (restart)'
    end
    sleep(1.5)
    run_app
    res.body = IO::read("#{File.dirname(__FILE__)}/ruined/html/waitclient.html")
  end
  svr.mount_proc('/quit') do |req, res|
    begin
      kill_child
    rescue
      svr.logger.error 'failed to kill child (quit)'
    end
    c = 0
    if req.path =~ %r|/(\d+)|
      c = $1.to_i
    end
    quit_svr(c, svr)
    res.body = '<html>bye</html>'
  end
  svr.mount_proc('/connect') do |req, res|
    begin
      open('http://127.0.0.1:8383/debug/start') do |http|
        http.read
      end
    rescue
      res.status = 404
    end
    res.body = '<html></html>'
  end

  Thread.start do
    run_app
    if RUBY_PLATFORM =~ /win32/
      system('start http://localhost.:8384/html/index.html')
    elsif RUBY_PLATFORM =~ /cygwin/  
      system('cygstart http://localhost.:8384/html/index.html')      
    else
      unless system('open', 'http://localhost.:8384/html/index.html')
        $stderr.puts 'open http://localhost:8384/ on your web browser'
      end
    end
  end
  svr.start
  puts 'debugger exit'
  exit(0)
end


lib/ruined/ruinmain.rb
#!/usr/local/bin/ruby
# coding: utf-8

require 'webrick'
require 'uri'
require 'json'
require 'thread'
require 'monitor'
require 'stringio'
require 'fileutils'

module Ruined
  RUINED_VERSION = '0.1.0'
  
  @queue = [Queue.new, Queue.new]
  @breakpoints = []
  @monitor = Monitor.new
  IGNORES = [:$&, :$', :$+, :$_, :$`, :$~, :$KCODE, :$= ]
  @unbreakable_threads = []
  @user_threads = { }

  class <<Thread
    alias :_original_start :start
    def start(&proc)
      webrick = caller.first.include?('webrick')
      $stderr.puts "caller=#{caller[2]}, webrick=#{webrick}"
      $stderr.flush
      _original_start do
        Ruined.add_unbreakable(Thread.current) if webrick
        begin
          proc.call
        ensure
          Ruined.remove_unbreakable(Thread.current) if webrick    
        end
      end  
    end
  end
  
  class Context
    TLSES = ['$!', '$?', '$@', '$SAFE']
    def initialize(e, f, l, id, bnd, b, s)
      @event = e
      @file = f
      @line = l
      @iid = id.to_s
      @break = b
      @binding = bnd
      @stdout = s
      @tlses = Hash[*(TLSES.map{|k| [k, eval(k)]}.flatten(1))]
    end
    def to_hash
      { :event => @event, :file => @file, :line => @line, :id => @iid, :break => @break,
        :stdout => @stdout, :threads => Ruined.user_threads(self) }
    end
    attr_reader :tlses, :file, :line, :iid, :binding, :break
    attr_accessor :event, :stdout
  end

  include WEBrick
  svr = HTTPServer.new(:Port => 8383,
                       :ServerType => Thread,
                       :Logger => ($DEBUG) ? Log.new(nil, BasicLog::DEBUG) : Log.new,
                       :DocumentRoot => File.dirname(__FILE__))
  trap('INT') do 
    svr.shutdown
  end
  
  class DebugServlet <  HTTPServlet::AbstractServlet
    include WEBrick::HTMLUtils
    def service(req, res)
      if Ruined.local_call?(req.addr)
        @response = res
        @request = req
        super
        @response = nil
        @request = nil
      else
        bye(res)
      end
    end
    def do_GET(req, res)
      m = %r|/debug/([^/?]+)/?([^?]*).*\Z|.match(req.unparsed_uri)
      if m
        res.body = __send__(m[1].to_sym, *(m[2].split('/').map{|x|URI.decode(x)}))
      else
        bye(res)        
      end
    end
    def do_POST(req, res)
      m = %r|/debug/([^/?]+)/?([^?]*).*\Z|.match(req.unparsed_uri)
      if m && m[1] == 'file'
        res.body = __send__(m[1].to_sym, *(m[2].split('/').map{|x|URI.decode(x)}))
      else
        bye(res)        
      end
    end

    attr_reader :response, :request

    def break(*a)
      if a.size < 3
        bye(response)
      else
        point = [a[1..(a.size - 2)].join('/'), a[a.size - 1].to_i]
        if a[0] == 'true'
          Ruined.breakpoints << point
        else
          Ruined.breakpoints.delete point
        end
        JSON(point)
      end
    end

    def run(*a)
    end

    def stop(*a)
    end

    def stepping(*a)
      Ruined.wait 1      
      JSON(Ruined.current_context)
    end

    def cont(*a)
      Ruined.release 0
      Ruined.wait 1
      JSON(Ruined.current_context)
    end

    def step(*a)
      cont(a)
    end
        
    def file(*a)
      file = a.join('/')
      if request.request_method == 'GET'
        request.accept.each do |accept|
          if accept =~ /html/i
            break
          elsif accept =~ %r|text/plain|i
            File.open(file) do |f|
              return Ruined::to_utf8(f.read)
            end
          end
        end
        to_html(file)
      else
        # TODO: newline and charset justify
        # wb is workaround for inhibit to put addition \r
        File.open("#{file}.new", 'wb') do |f|
          f.write(request.body)
        end
        FileUtils.cp file, "#{file}.bak", :verbose => true
        FileUtils.cp "#{file}.new", file, :verbose => true
        FileUtils.rm "#{file}.new"
        "save #{file}"
      end  
    end
    
    def locals(*a)
      if a.size == 0
        create_varlist Ruined.local_vars
      else
        eval_var(a)
      end
    end
    
    def globals(*a)
      if a.size == 0
        create_varlist Ruined.global_vars
      else
        eval_var(a)
      end
    end

    def self(*a)
      if a.size == 0
        create_varlist Ruined.self_vars
      elsif a.size < 2
        bye(response)
      else
        eval_var(a)
      end
    end
    
    def start(*a)
      '<html>start</html>'
    end
    
    private
    
    def create_varlist(t)
      s = '<table class="vars"><tr><th>Name</th><th>Value</th></tr>'
      t.each do |e|
        v = Ruined.to_utf8(e[:value].inspect)
        s << "<tr><td>#{e[:name]}</td><td class=\"var-value\">#{escape(v)}</td></tr>"
      end
      s + '</table>'
    end

    def eval_var(a)
      if a.size < 2
        bye(response)
      else
        Ruined.set(a[0], a[1..-1].join('/')).to_s
      end
    end
    
    def to_html(file)
      r = '<table>'
      File.open(file).each_line do |line|
        r << "<tr><td><pre>#{escape(Ruined::to_utf8(line))}</pre></td></tr>"
      end.close
      r + '</table>'
    end
    
    def bye(res)
      res.status = 404
      res.body = '<html>bye</html>'
    end
  end

  def self.current_context
    raise RuntimeException.new('bug: no context !!') unless @current
    @current.to_hash
  end
  
  def self.breakpoints
    @breakpoints
  end
  
  def self.local_vars
    script = <<EOD
local_variables.map do |_0|
  (_0 == :_) ? nil : { :name => _0.to_s, :value => eval(_0.to_s) }
end - [nil]
EOD
    (@current) ? eval(script, @current.binding) : []
  end
  
  def self.self_vars
    script = <<EOD
[{ :name => 'class', :value => self.class.to_s }] +
instance_variables.map do |v|
  { :name => v.to_s, :value => instance_eval(v.to_s) }
end +
self.class.class_variables.map do |v|
  { :name => v.to_s, :value => instance_eval(v.to_s) }
end
EOD
    (@current) ? eval(script, @current.binding) : []    
  end
  
  def self.global_vars
    script = <<EOD
(global_variables - Ruined::IGNORES).map do |v|
  if v.to_s =~ /\\A\\$[1-9]/
    nil
  else
    { :name => v.to_s, :value => eval(v.to_s) }
  end
end - [nil]
EOD
    a = eval(script)
    if @current
      0.upto(a.size - 1) do |i|
        if @current.tlses.has_key?(a[i][:name])
          a[i][:value] = @current.tlses[a[i][:name]]
        end
      end
    end
    a
  end
  
  def self.set(var, val)
    eval("#{var} = #{val}", @current.binding)
  end
  
  def self.wait(t)
    return unless @queue    
    logger.debug("------------wait #{t}")
    o = @queue[t].pop
    if t == 1
      @current = o
    end
    logger.debug("------------wait exit #{t}")
  end

  def self.release(t, obj = nil)
    return unless @queue
    logger.debug("------------release #{t}")
    @monitor.synchronize {    
      @queue[t].push obj
    }
    logger.debug("------------release exit #{t}")
  end
  
  def self.output
    return '' unless StringIO === $stdout
    out = $stdout
    $stdout = StringIO.new
    out.pos = 0
    ret = ''
    out.each_line do |x|
      ret << "#{HTMLUtils.escape(to_utf8(x.chomp))}<br/>"
    end
    ret
  end
  
  def self.to_utf8(s)
    (s.encoding != Encoding::UTF_8) ? s.encode(Encoding::UTF_8) : s
  end
  
  def self.add_unbreakable(t)
    @monitor.synchronize {
      @unbreakable_threads << t
    }
  end
  
  def self.remove_unbreakable(t)
    @monitor.synchronize {
      @unbreakable_threads.delete t
    }
  end
  
  def self.unbreakable?(t)
    @monitor.synchronize {
      @unbreakable_threads.include? t
    }
  end
  
  def self.local_call?(addr)
    ['127.0.0.1', '::1'].include?(addr[3])
  end
  
  def self.user_threads(current)
    r = []
    @monitor.synchronize {                
      @user_threads.each do |t, c|
        r << [:file => File.basename(c.file), :line => c.line, 
              :self => (c == current),
              :status => status_to_s(t.status)]
      end
      r
    }
  end
  
  def self.status_to_s(s)
    if s.nil?
      'aborted'
    elsif s == false
      'dead'
    else
      s
    end
  end

  svr.mount('/debug', DebugServlet)
  svr.mount_proc('/quit') do |req, res|
    if local_call?(req.addr)
      set_trace_func(nil)
      c = 0
      if req.path =~ %r|/(\d+)|
        c = $1.to_i
      end
      res.body = '<html>bye</html>'
      Thread.start do
        @monitor.synchronize {        
          @queue = nil
        }
        Thread.pass
        svr.shutdown
        exit(c)
      end
    else
      res.status = 404
    end
  end
  
  define_method(:logger) do
    return svr.logger
  end
  module_function(:logger)

  svr.start

  set_trace_func Proc.new {|event, file, line, id, binding, klass|
    unless file =~ %r#(lib/ruby|ruinmain|webrick|internal)# || unbreakable?(Thread.current)
      if event.index('c-') != 0
        if file == $0 && !$stdout.instance_of?(StringIO)
          $stdout = StringIO.new
        end
        b = breakpoints.include? [file, line]
        ctxt = @monitor.synchronize {            
          @user_threads[Thread.current] = Context.new(event, file, line, id, binding, b, output)
        }
        svr.logger.debug(ctxt.inspect)
        release 1, ctxt
        wait 0
        svr.logger.debug('continue...')
      end
    end
  }
  at_exit { 
    if @current
      @current.event = 'exit'
      @current.stdout = output
      release 1, @current #reschedule
      wait 0
    end
  }
end
